numpy/core/src/umath/ufunc_object.c

```diff
@@ -3613,46 +3613,76 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
         return NULL;
     }
 
-    /* Get the identity */
-    identity = _get_identity(ufunc, &reorderable);
-    if (identity == NULL) {
+    /* Get the reduction dtype */
+    if (reduce_type_resolver(ufunc, arr, odtype, &dtype) < 0) {
+        Py_DECREF(initial);
         return NULL;
     }
 
+
+    PyArrayObject *initial_arr = NULL;
+
     /* Get the initial value */
     if (initial == NULL || initial == NoValue) {
-        initial = identity;
-
+        PyObject *identity;
         /*
         * The identity for a dynamic dtype like
         * object arrays can't be used in general
         */
-        if (initial != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
-            Py_DECREF(initial);
-            initial = Py_None;
-            Py_INCREF(initial);
+        if (PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
+            identity = Py_None;
+            Py_INCREF(identity);
+        }
+        else {
+            /* Get the identity */
+            identity = _get_identity(ufunc, &reorderable);
+            if (identity == NULL) {
+                return NULL;
+            }
+        }
+
+        if (identity == Py_None) {
+            initial_arr = NULL;
+        }
+        else {
+            /* upcast to a 0d array of the right type, with deliberately vague casting */
+            initial_arr = PyArray_NewFromDescr(
+                &PyArray_Type,
+                dtype,
+                /* zero-dimensional */
+                0, NULL, NULL,
+                NULL, 0,NULL);
+            if (initial_arr == NULL) {
+                return NULL;
+            }
+            if (PyArray_FillWithScalar(initial_arr, identity) < 0) {
+                return NULL;
+            }
         }
-    } else {
         Py_DECREF(identity);
-        Py_INCREF(initial);  /* match the reference count in the if above */
     }
-
-    /* Get the reduction dtype */
-    if (reduce_type_resolver(ufunc, arr, odtype, &dtype) < 0) {
-        Py_DECREF(initial);
-        return NULL;
+    else if (initial == Py_None) {
+        /* identity explicitly disallowed */
+        initial_arr = NULL;
+    }
+    else {
+        /* infer the dtype, ReduceWrapper will perform same_kind casting to check later */
+        initial_arr = (PyArrayObject *)PyArray_FROM_O(initial);
```

Comment:
I don't remember if `PyArray_FROM_O` makes a copy, but my intent was that it not do so.